solr-ingest/src/main/java/edu/unc/lib/dl/data/ingest/solr/filter/SetDatastreamFilter.java

```diff
@@ -0,0 +1,145 @@
+        List<Datastream> datastreams = getDatastreams(fileObj, ownedByOtherObject);
+
+        IndexDocumentBean doc = dip.getDocument();
+
+        doc.setDatastream(getDatastreamStrings(datastreams));
+        doc.setFilesizeTotal(getFilesizeTotal(datastreams));
+        doc.setFilesizeSort(getFilesize(datastreams));
+    }
+
+    private FileObject getFileObject(ContentObject contentObj) {
+        if (contentObj instanceof FileObject) {
+            return (FileObject) contentObj;
+        } else if (contentObj instanceof WorkObject) {
+            WorkObject workObj = (WorkObject) contentObj;
+            return workObj.getPrimaryObject();
+        } else {
+            // object being indexed must be a work or a file object
+            return null;
+        }
+    }
+
+    /**
+     * Generates a list of Datastream objects from each binary belonging to the
+     * provided FileObject. If the datastreams are being recorded on an object
+     * other than their owning file object, the pid of the owning file object is
+     * recorded
+     * 
+     * @param fileObj
+     * @param ownedByOtherObject
+     * @return
+     */
+    private List<Datastream> getDatastreams(FileObject fileObj, boolean ownedByOtherObject) {
+        return fileObj.getBinaryObjects().stream()
+            .map(binary -> {
+                Resource binaryResc = binary.getResource();
+
+                String name = binaryResc.getURI();
+                name = name.substring(name.lastIndexOf('/') + 1);
+
+                String mimetype = binaryResc.getProperty(Ebucore.hasMimeType).getString();
+                long filesize = binaryResc.getProperty(Premis.hasSize).getLong();
+                // Making assumption that there is only one checksum
+                String checksum = binaryResc.getProperty(Premis.hasMessageDigest).getResource().getURI();
+
+                String filename = binaryResc.getProperty(Ebucore.filename).getString();
+                int extensionIndex = filename.lastIndexOf('.');
+                String extension = extensionIndex == -1 ? "" : filename.substring(extensionIndex + 1);
+
+                String owner = ownedByOtherObject ? fileObj.getPid().getId() : null;
+
+                return new Datastream(owner, name, filesize, mimetype, filename, extension, checksum);
+            }).collect(Collectors.toList());
+    }
+
+    private List<String> getDatastreamStrings(List<Datastream> datastreams) {
+        return datastreams.stream()
+                .map(Datastream::toString)
+                .collect(Collectors.toList());
+    }
+
+    /**
+     * Returns the sum of filesizes for all datastreams which do no belong to
+     * other objects
+     * 
+     * @param datastreams
+     * @return
+     */
+    private long getFilesizeTotal(List<Datastream> datastreams) {
+        return datastreams.stream()
+            .filter(ds -> ds.getFilesize() != null && ds.getOwner() == null)
```

Comment:
should this be ds.getFilesize() != 0 instead? getFilesize() shouldn't return null, right?